Fix decoding empty Cloudflare SRV target (plus fixes for a few other providers)

geek1011 opened the pull request:
SRV records with empty (".") targets are now returned as false by the API, which breaks Unmarshaling it into a string.

fixes #561

geek1011 commented on integrationTest/integration_test.go:
```diff
@@ -458,6 +458,7 @@ func makeTests(t *testing.T) []*TestCase {
 			tc("Change Priority", srv("_sip._tcp", 52, 6, 7, "foo.com."), srv("_sip._tcp", 15, 65, 75, "foo4.com.")),
 			tc("Change Weight", srv("_sip._tcp", 52, 62, 7, "foo.com."), srv("_sip._tcp", 15, 65, 75, "foo4.com.")),
 			tc("Change Port", srv("_sip._tcp", 52, 62, 72, "foo.com."), srv("_sip._tcp", 15, 65, 75, "foo4.com.")),
+			tc("Null Target", srv("_sip._tcp", 52, 62, 72, "foo.com."), srv("_sip._tcp", 15, 65, 75, ".")),
```
This may break other providers if they don't handle empty SRV records properly.

tlimoncelli replied:
Did you see this in the wild?  It seems like a bad thing to do, even if it is legal.  Like, maybe we should add a new opinion that bans this kind of thing (not because it is illegal, but because it is ill-advised).

geek1011 replied:
Yes, I've seen it in the wild (both its usage and providers not supporting it, I don't know which you're referring to). For example, [Fastmail](https://www.fastmail.com/help/receive/domains-advanced.html) and [Gmail](https://dns.google.com/resolve?name=_pop3._tcp.gmail.com&type=SRV) both use this for pop3. I've also seen it for other services like unencrypted SIP ports on VoIP providers, but I don't recall exactly which ones at the moment. For providers not supporting it, I have heard Plesk didn't support it.

tlimoncelli replied:
I shouldn't read Gihub from my phone. I totally missed that "." is a legal target (you had written about that above). My apologies.

geek1011 replied:
No worries.

geek1011 commented on providers/cloudflare/cloudflareProvider.go:
```diff
@@ -422,23 +423,46 @@ func newCloudflare(m map[string]string, metadata json.RawMessage) (providers.DNS
+		return v
+	case bool:
+		if !v {
+			return "" // the "." is already added by nativeToRecord
```
This is a slightly ugly hack, and will need to be kept in mind if ever using `Target()` in other places. To actually do this properly, nativeToRecord shouldn't add a `.` if it's already a FQDN. I'll fix this if it doesn't cause other side effects (I'm not too familiar with the cloudflare provider code yet).

geek1011 replied:
I've decided to leave this as-is, as the alternative would be to add the `.` directly in this function, which will either greatly increase the complexity of `Target()` and `GetTarget()` (checking for the trailing `.`, etc) or would cause them not to be the inverse of each other in all cases (which could introduce obscure bugs).

geek1011 commented:
I've tested this, and it works for me. I haven't ran the integration tests, though.

captncraig commented:
🎱

captncraig commented:
/AzurePipelines run Integration Tests

geek1011 commented:
So, it seems the following needs to be fixed:
- Type for sending modification to Cloudflare (decoding works, encoding is slightly incorrect)
- DigitalOcean has the double-dot issue (similar to what I was pointing out previously)
- NameDotCom doesn't support null targets
- DNSimple returns a null record in an nonstandard format (the target is omitted entirely)

I'll fix CF and DO later today.

geek1011 commented:
Ok, I've pushed a rewrite of this PR which should make CF work properly in all cases. In addition, it's more resilient to future API changes and it makes the AddOrigin usage less hacky.

geek1011 commented:
I've fixed DO, and I think I've fixed DNSimple (I can't test it as I don't have an account).

@captncraig, can you run the integration tests again?

@tlimoncelli, what do you want to do about NameDotCom? Would you prefer to skip the test for it, or to add another capability for null SRV targets?

captncraig commented:
If the provider fails on a technically correct case like this, generally I will handle it in the provider code and return an error. In this case I'd probably have namedotcom return a hard error, as opposed to printing a warning, since it isn't possible to create the requested config. Make sure the message is somewhat descriptive of what the problem is.

captncraig commented:
🎱

captncraig commented:
/AzurePipelines run Integration Tests

geek1011 commented:
DNSImple now seems to be failing with `panic: unparsable record received from dnsimple: SRV value too big for uint16: strconv.ParseUint: parsing "75.": invalid syntax [recovered]`. I don't know why the 75 has a . after it. Can someone with access to the DNSimple API post the response for a SRV record with an empty target so I can have a look?

geek1011 commented:
I've now fixed everything now except for the last DNSimple error, which I'll need to see the API output for.

tlimoncelli commented:
@aeden can you assist with testing DNSimple?

onlyhavecans commented:
I am looking into the DNSimple breakage and will try to have a patch tomorrow

geek1011 commented:
@onlyhavecans, any update on this?

onlyhavecans commented:
Hi! This is not an API but a logic bug.  I opened a PR against your repo @geek1011

geek1011 commented:
@captncraig, can you run integration tests again?

captncraig commented:
/AzurePipelines run Integration Tests

geek1011 commented:
I'm finished with this PR. All tests pass now (GCloud failed in the last run because someone seems to have removed the domain from the test account, but it passed last time, and nothing has changed since then).

tlimoncelli commented:
Awesome!  Thanks for this contribution!
